migen/build/lattice/icestorm.py

```diff
@@ -32,31 +32,51 @@ def _build_pcf(named_sc, named_pc):
     return r
 
 
-def _build_yosys(device, sources, vincpaths, build_name):
-    ys_contents = ""
+def _build_yosys(device, sources, vincpaths, build_name, pre_synth,
+                 synth_override, post_synth):
+    # Unconditionally-required yosys commands in the context of Migen.
+    ys_contents = list()
     incflags = ""
     for path in vincpaths:
         incflags += " -I" + path
     for filename, language, library in sources:
-        ys_contents += "read_{}{} {}\n".format(language, incflags, filename)
+        ys_contents.append("read_{}{} {}".format(language, incflags, filename))
+
+    # Migen only outputs Xilinx-style attributes enclosed in strings.
+    # (i.e. "true", "0", etc). Yosys wants constant literals to represent
+    # true and false, so convert before synthesis.
+    ys_contents.append("attrmap -tocase keep -imap keep=\"true\" keep=1 -imap keep=\"false\" keep=0 -remove keep=0")
+
+    # Pre-synthesis user-commands go here.
+    for pre in pre_synth:
+        ys_contents.append(pre.format(build_name=build_name))
+
+    # Give the programmer an opportunity to override the default synthesis
+    # command if they need to. The default is fine for most applications.
+    # Must generate {build_name}.blif as output.
+    if not synth_override:
+        ys_contents.append("""synth_ice40 -top top -blif {build_name}.blif""".format(
+            build_name=build_name))
+    else:
+        for synth in synth_override:
+            ys_contents.append(synth.format(build_name=build_name))
 
-    ys_contents += """synth_ice40 -top top -blif {build_name}.blif""".format(
-        build_name=build_name)
+    # Finally, add post-synthesis commands.
+    for post in post_synth:
+        ys_contents.append(post.format(build_name=build_name))
 
     ys_name = build_name + ".ys"
-    tools.write_to_file(ys_name, ys_contents)
+    tools.write_to_file(ys_name, "\n".join(ys_contents))
```

Comment:
Merge `pre_synth`, `post_synth`, `synth_override`, the `"read_..."` and `"attrmap ..."` all into one list of format strings that can be modified by platforms and is then formatted and output.